utils/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/source/UniversalKafkaSourceFactory.scala

```diff
@@ -281,6 +279,74 @@ class UniversalKafkaSourceFactory(
     )
   }
 
+  @transient protected lazy val dataSampleParamName: ParameterName =
+    KafkaUniversalComponentTransformer.dataSampleParamName
+
+  private def afterSchemaParamStep(
+      nextParams: List[Parameter]
+  ): ContextTransformationDefinition = {
+    case TransformationStep(
+          (`topicParamName`, DefinedEagerParameter(_: String, _)) ::
+          (`contentTypeParamName`, DefinedEagerParameter(contentType: String, _)) :: Nil,
+          _
+        ) if contentType == ContentTypes.JSON.toString && enableSchemaDerivationFromDataSampleForSchemalessJsonTopics =>
+      val dataSampleParam = getJsonDataSampleParam
+      NextParameters(parameters = dataSampleParam.createParameter() :: nextParams)
+    case TransformationStep(
+          (`topicParamName`, DefinedEagerParameter(_: String, _)) ::
+          (`contentTypeParamName`, DefinedEagerParameter(_: String, _)) :: Nil,
+          _
+        ) if nextParams.nonEmpty =>
+      NextParameters(parameters = nextParams)
+    case TransformationStep(
+          (`topicParamName`, DefinedEagerParameter(_: String, _)) ::
+          (`schemaVersionParamName`, DefinedEagerParameter(_: String, _)) :: Nil,
+          _
+        ) if nextParams.nonEmpty =>
+      NextParameters(parameters = nextParams)
+  }
+
+  private def getJsonDataSampleParam: ParameterCreatorWithNoDependency with ParameterExtractor[Any] = {
+    ParameterDeclaration
+      .optional[Any](dataSampleParamName)
+      .withCreator(
+        modify = _.copy(
+          editors = List(JsonParameterEditor),
+          defaultValue = Some(Expression.json("{}")),
+          hintText = Some(
+            "Provide an example JSON data sample. It will be analyzed to determine field types and generate a schema for easier data access in the subsequent nodes."
+          ),
+        )
+      )
+  }
+
+  private def runtimeDataForPlainSchema = {
+    RuntimeSchemaData[ParsedSchema](
+      new NkSerializableParsedSchema[ParsedSchema](ContentTypesSchemas.schemaForPlain),
+      Some(SchemaId.fromString(ContentTypes.PLAIN.toString))
+    )
+  }
+
+  private def runtimeDataForJsonSchema = {
+    RuntimeSchemaData[ParsedSchema](
+      new NkSerializableParsedSchema[ParsedSchema](ContentTypesSchemas.schemaForJson),
+      Some(SchemaId.fromString(ContentTypes.JSON.toString))
+    )
+  }
+
+  private def typingResultWithoutValues(typingResult: TypingResult) = {
+    typingResult.withoutValue match {
+      case obj: TypedObjectTypingResult if obj.fields.isEmpty => Unknown
```

Comment:
Let's add a comment that it is a kind of heuristic - we assume that user passes `{}` when the are not sure what will be in fields.
Please extract this method to some new class e.g. `TypingResultFromJsonTypeDeterminer`

Reply:
I'd rather thinking about some closer place to this class (e.g. in the same module, next to this class). I want to keep this logic easy to reuse but not too make it too much generic yet because it can be not easy to figure out that it is a type base on example json. BTW proposed by name doesn't include `Example` which is very important in this context. Maybe let's call it `TypingResultFromJsonExampleTypeDeterminer`.